README.md
[![Code
Climate](https://codeclimate.com/github/AutomatedInsightsInc/wordsmith-ruby-sdk/badges/gpa.svg)](https://codeclimate.com/github/AutomatedInsightsInc/wordsmith-ruby-sdk)
[![Test
Coverage](https://codeclimate.com/github/AutomatedInsightsInc/wordsmith-ruby-sdk/badges/coverage.svg)](https://codeclimate.com/github/AutomatedInsightsInc/wordsmith-ruby-sdk/coverage)
[![Issue
Count](https://codeclimate.com/github/AutomatedInsightsInc/wordsmith-ruby-sdk/badges/issue_count.svg)](https://codeclimate.com/github/AutomatedInsightsInc/wordsmith-ruby-sdk)

# Wordsmith SDK for Ruby
## Intro to Wordsmith

[Wordsmith](http://wordsmith.automatedinsights.com) makes it easy to generate thousands of stories, reports, and articles
in the time it takes to write just one. Wordsmith is a natural language
generation tool that enables users to turn data into text using dynamic
templates. The platform is easy to learn, and powerful enough to make each piece
of content totally unique.

The Wordsmith API allows developers to generate new content using the Templates
created in the Wordsmith web app by users at your company. Developers can use
the API to send new data which will trigger the generation of new text content.
You have complete control over when you generate and use the content.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'wordsmith-ruby-sdk'
```

And then execute:

    $ bundle install


## Configuration
Configure in initializer or as desired

```ruby
#config/initializers/wordsmith.rb
Wordsmith.configure do |config|
  config.token = MY_API_TOKEN
  config.url = 'https://api.automatedinsights.com/v1' #optional, this is the default value
end
```
  
## Usage
```ruby
projects = Wordsmith::Project.all                 #an array of projects your token can access
project = Wordsmith::Project.find('project-slug') #fetch a project by slug

project.schema     #the data schema for the project
project.templates  #a collection of templates for this project

template = project.templates.find('template-slug')           #fetch a template by slug
template.generate({a_data_point: 1, another_one: 'Tuesday'}) #generate content
 > {content: 'Your content is here!'}
```

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/automatedinsightsinc/wordsmith-ruby-sdk.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](https://github.com/automatedinsightsinc/wordsmith-ruby-sdk/compare/v1.0.5...HEAD)

##### Changed
## [1.0.5](https://github.com/automatedinsightsinc/wordsmith-ruby-sdk/compare/v1.0.4...v1.0.5)
##### Changed
- Return :errors from Wordsmith when HTTP response code is 400 

## [1.0.4](https://github.com/automatedinsightsinc/wordsmith-ruby-sdk/compare/v1.0.3...v1.0.4)
##### Added
- Handled HTTP 429

## [1.0.3](https://github.com/automatedinsightsinc/wordsmith-ruby-sdk/compare/v1.0.2...v1.0.3)
##### Changed
- Updated for API v1.1
- Fix to handle additional fields returned by the API.

## [1.0.2](https://github.com/AutomatedInsightsInc/wordsmith-ruby-sdk/tree/v1.0.2) - 2016-04-27
##### Added
- Codeclimate test coverage reporter with dotenv for codeclimate token.
- Required ruby version in gemspec.
- Gem version number bumped (missed bump in 1.0.0 release)

## 1.0.0 - 2015-03-30

##### Added
- Enabled test converage for generating output since encoding has been fixed on the api.
- A place to put a test token in `test/test_helper.rb`, should be Robert's token to match actual projects to test against.

##### Changed
- Updated configuration default url to https://api.automatedinsights.com/v1
- Updated project generate output url to match routes in v1 api.

## 0.0.2 - 2015-02-02

##### Added
- Test coverage using minitest.
- Lots of README updates.
- CHANGELOG file.

##### Removed
- Rails dependencies.
- Configuration through yaml file.

##### Changed
- Configuration through conventional configuration methods.

## 0.0.1 - 2015-01-21
- Prototype requiring railties.


test/client_test.rb
require 'test_helper'

class ClientTest < Minitest::Test
#  def test_get
#    skip
#  end
#
#  def test_post
#    skip
#  end
end

class ClientTest::ParseResponseTest < Minitest::Test
end

class ClientTest::ConnectionTest < Minitest::Test
  def setup
    Wordsmith.configure do |config|
      config.token = 'some_token'
      config.url = 'http://www.some-url.com/api'
    end
    super
  end

  def teardown
    Wordsmith.reset
    super
  end

  def test_url_configuration_change
    connection = Wordsmith.client.send :connection
    assert_equal 'http://www.some-url.com/api',
      connection.url_prefix.to_s

    Wordsmith.configure do |config|
      config.url = 'http://www.some-other-url.com/api'
    end

    connection = Wordsmith.client.send :connection
    assert_equal 'http://www.some-other-url.com/api',
      connection.url_prefix.to_s
  end

  def test_token_configuration_change
    connection = Wordsmith.client.send :connection
    assert_equal 'Bearer some_token',
      connection.headers['Authorization']

    Wordsmith.configure do |config|
      config.token = 'some_other_token'
    end

    connection = Wordsmith.client.send :connection
    assert_equal 'Bearer some_other_token',
      connection.headers['Authorization']
  end

  def test_content_type_header_is_json
    connection = Wordsmith.client.send :connection
    content_type = connection.headers['Content-Type']
    assert_equal 'application/json', content_type
  end

  def test_authorization_header
    connection = Wordsmith.client.send :connection
    authorization = connection.headers['Authorization']
    assert_equal 'Bearer some_token', authorization
  end
end


test/test_helper.rb
WORDSMITH_SDK_TEST_TOKEN='923b278a6088675262af64ceb437bab31d7ebc6b07aaf89f88b0b88dd4fe2a97'

require 'dotenv'
Dotenv.load

if ENV['CODECLIMATE_REPO_TOKEN']
  require 'codeclimate-test-reporter'
  CodeClimate::TestReporter.start
end

require 'wordsmith-ruby-sdk'
require 'minitest/autorun'
require 'minitest/unit'
require 'minitest/pride'


test/project_test.rb
require 'test_helper'

class ProjectTest < Minitest::Test
  def setup
    Wordsmith.configure do |config|
      config.token = WORDSMITH_SDK_TEST_TOKEN
    end
  end

  def teardown
    Wordsmith.reset
  end

  def test_all_contains_project
    list = Wordsmith::Project.all
    assert list.find {|p| p.slug == 'test'}
  end

  def test_find_returns_project
    project = Wordsmith::Project.find 'test'
    assert_instance_of Wordsmith::Project, project
  end

  def test_find_raises_if_not_found
    err = assert_raises RuntimeError do
      Wordsmith::Project.find 'test-sdk-nonexistent-project'
    end

    assert_equal %Q(Project not found with slug: "test-sdk-nonexistent-project"),
      err.message
  end

  def test_schema
    project = Wordsmith::Project.find 'test'
    expected = {a: 'Number', b: 'Number', c: 'Number'}
    assert_equal expected, project.schema
  end

  def test_template_collection_exists
    project = Wordsmith::Project.find 'test'
    assert_instance_of Wordsmith::TemplateCollection, project.templates
  end
end


lib/wordsmith/client.rb
require 'faraday'
require 'json'
require 'hashie'

module Wordsmith
  class Client

    def get(uri)
      response = connection.get uri
      parse_response response
    end

    def post(uri, data)
      response = connection.post uri, {data: data}.to_json
      parse_response response
    end

    private

    def connection
      return @_connection if connection_valid?
      @_connection = Faraday.new(url: Wordsmith.configuration.url, headers: {
        'Content-Type' => 'application/json',
        'Authorization' => "Bearer #{Wordsmith.configuration.token}",
        'User-Agent' => Wordsmith.configuration.user_agent
      })
    end

    def connection_valid?
      return false unless @_connection
      url = @_connection.url_prefix.to_s
      authorization = @_connection.headers['Authorization']

      url == Wordsmith.configuration.url &&
        authorization == "Bearer #{Wordsmith.configuration.token}"
    end

    def parse_response(response)
      body = JSON.parse(response.body)
      Hashie.symbolize_keys! body
      case response.status
      when 200, 201
        return body[:data]
      when 400
        fail %Q(Bad Request: "#{body[:errors]}")
      when 401
        fail 'API authorization error'
      when 429
        fail body[:error]
      else
        fail 'API error'
      end
    end
  end

  module_function

  def client
    @_client ||= Client.new
  end
end
